(Closes #2125) Fix bugs in SymbolTable deep copy and InlineTrans when array bounds reference formal arg

arporter opened the pull request:
Fixes bugs in SymbolTable deep copy for for symbol initial values and shape specification. Also fixes a bug in InlineTrans when the lower bound of an array formal argument is specified by a reference to another formal argument.
Finally, adds the `force` option to InlineTrans to permit it to ignore CodeBlocks within the target routine.

arporter commented:
This is ready for a first look now. One for @sergisiso, @hiker or @LonelyCat124. I've triggered the integration tests.

arporter commented:
(I am wondering whether having the "force" option to InlineTrans mean "ignore CodeBlocks" is a good idea: perhaps it should be more specific such as "permit-codeblocks"?)

sergisiso commented on src/psyclone/psyir/symbols/symbol_table.py:
```diff
@@ -303,6 +302,40 @@ def deep_copy(self):
                                      routine.from_container))
             symbol.routines = new_routines
 
+        # Ensure any Symbols referenced in array shapes are also updated
```
In fact, not for this PR but this method may also be useful for nodes to reattach them to scopes with different symbols by using the current names. Right now we just reattach thing hoping that symbols are there without any guarantees.

arporter replied:
Yes, it might also be useful/re-usable in inlining. I've done your `relink` suggestion and it works out quite nicely (although I had to add it to `Node` too). Currently the `Node.relink` uses `walk` to find References but obviously this will miss (at least) precision symbols in Literals. I remember we've got at least one TODO to switch to using reference_accesses instead to avoid this problem. Do you think I should do that here? Also, I'm thinking that perhaps `replace_symbols` or `update_symbols` might be a more self-explanatory name rather than `relink`. What do you think?

arporter replied:
The issue for reference_accesses is #2542 and that depends on #2424 (make it work for nested scopes). That's not going to happen any time soon I think (I previously looked at doing #2424 and it got very difficult). There are a few possible approaches to replacing Symbols in a PSyIR tree:
 * Add a `relink` method to Node and have it look for all References (and Literals) and replace their symbols;
 * Add a `relink` method to Node that simply walks the tree calling `relink`. Then implement the necessary functionality in `relink` methods for `Reference` and `Literal` classes (at least).
 * Create a Visitor with handlers for `Reference`, `Literal`, `Routine`...

sergisiso replied:
I do not mind the name, in fact linking is probably too broad for Node, as there are several "links". Also for Node it does not need a symbol_table parameter, as to avoid the same problems with nested scopes it should you a lookup from the scope it is connected in. Maybe "update_symbols_to_current_scope" or "bind_symbols_to_current_scope"? In symbols it can be "bind_symbols_to(symbol_table)", or something similar?

Regarding the implementation, I see you started with the Visitor, but I am unsure about this choice. A Visitor still decouples the implementation for node, which is good for having a multiple backends but when adding mandatory functionality that depends on the attributes of a class, I think its clear if its all in the same place, with a specialised method in this class (your second bullet point). This way, when some body adds or removes an attribute from a node, it has to check if it needs to update the eq, copy, relink or that class (instead of knowing how other files may be affected). Also visitors do not do that well with inheritance. While a overwiting method only has to care about its attributes, and then calling the super() version, which will care about theirs. But I am happy to wait for the review to have a better opinion if you prefer the visitor.

sergisiso reviewed (changes requested):
@arporter I agree this needs to be done, as now we carelessly leave symbols pointing behind. But there are more symbols that the ones updated by this PR and the behaviour would be inconsistent if we update some but not others. Let me know if you want this PR sooner to fix a specific bug, but otherwise I suggest also updating all symbols inside symbols/datatypes.

arporter commented on src/psyclone/psyir/symbols/datatypes.py:
```diff
@@ -116,7 +124,7 @@ def __str__(self):
     def declaration(self):
         '''
-        :returns: the original declaration of the symbol. This is obviously \
+        :returns: the original declaration of the symbol. This is obviously
                   language specific and hence this class must be subclassed.
```
Gone.

arporter commented on src/psyclone/tests/psyir/symbols/datatype_test.py:
```diff
@@ -572,6 +591,27 @@ def test_arraytype_eq():
+    # But the new Reference should still refer to the same Symbol.
+    assert acopy.shape[0].upper.symbol is atype.shape[0].upper.symbol
+    assert acopy.shape[1].lower.symbol is atype.shape[1].lower.symbol
+    # When shape doesn't have set bounds.
```
Added and it does :-)

arporter commented on src/psyclone/psyir/symbols/datatypes.py:
```diff
@@ -709,6 +724,27 @@ def __eq__(self, other):
+        return ArrayType(self.datatype, new_shape)
+
 
 class StructureType(DataType):
```
Yep, you're quite right. Done now.

arporter commented on src/psyclone/psyir/symbols/datatypes.py:
```diff
@@ -709,6 +724,27 @@ def __eq__(self, other):
 
+    def copy(self):
+        '''
+        Create a shallow copy of this ArrayType. (Any References will be
+        re-created but the target Symbols will remain unchanged.)
```
Yes, done.

arporter commented on src/psyclone/psyir/symbols/datatypes.py:
```diff
@@ -116,7 +124,7 @@ def __str__(self):
+        :returns: the original declaration of the symbol. This is obviously
                   language specific and hence this class must be subclassed.
         :rtype: str
         '''
```
They weren't but they are now.

arporter commented on src/psyclone/psyir/symbols/datasymbol.py:
```diff
@@ -303,10 +303,21 @@ def copy(self):
 
         '''
         if self.initial_value is not None:
+            # Ensure any References in the initial-value expression are
```
Done and test added.

arporter commented:
Note to self: there's inevitably going to be missed coverage and I need to move quite a lot of testing out of `symboltable_test` and into individual tests of the various new `relink` methods.

arporter commented:
The visitor approach works out very neatly but I take your point about decoupling this functionality from the Nodes themselves. There's also quite a lot to be said for keeping the same method name for datatypes and nodes - it will make the code a bit clearer I think. Although it would be nice to remove the `table` argument for the Node method I don't think this is practical here as all the Nodes we care about are 'dangling' (in the `shape` of ArrayType or `initial_value` expressions of Symbols).

I'll try my 2nd bullet point then... :-)

arporter commented:
The `Node.update_symbols_from(table)` approach was almost as neat as the visitor so I've gone with it and renamed all the `relink` methods accordingly.

arporter commented:
Just need to tidy up/remove the SymbolTable tests now that functionality has been removed from the `deep_copy` method.

arporter commented:
I think this is ready for another look now. It has grown a lot so probably needs something of a fresh start.

sergisiso commented on src/psyclone/psyir/symbols/datasymbol.py:
```diff
@@ -327,3 +342,20 @@ def copy_properties(self, symbol_in):
         self._is_constant = symbol_in.is_constant
         self._initial_value = symbol_in.initial_value
+
+    def update_symbols_from(self, table):
```
Naming is hard  :) some more discussion before this makes it to master

I think the intent is more clear using the "replace" verb as in this docstring, update could mean many things and I think I wouldn't know what this method does without knowing the context. Could this be "replace_symbols_from"? also I am not sure about the preposition should it be "replace_symbols_with"? or "to"/"using"?

arporter replied:
I've gone for `replace_symbols_using`.

sergisiso commented on src/psyclone/psyir/nodes/reference.py:
```diff
@@ -250,6 +250,24 @@ def next_access(self):
             return all_accesses[index+1].node
         return None
 
+    def update_symbols_from(self, table):
```
Loop also has a symbol for its looping variable. We have a special case in "scoping_node.py::_refine_copy" for it. Could this responsibility be moved to a Loop "update_symbols_from" method?

arporter replied:
Great idea. Some code removed.

sergisiso reviewed (changes requested):
@arporter That's a big improvement, its satisfying to see that the resulting implementation of symbol_table::deep_copy is now very readable. The comments now are mostly about naming and documentation.

Could you add a section into https://psyclone-dev.readthedocs.io/en/latest/psyir_symbols.html regarding this. Essentially explaining that:
- The node, symbol, datatype copy operators generates new objects (including PSyIR subtrees when necessary, e.g. children, initial_values, precision, ...) but it keeps the references to the original symbols (so that the copied objects are valid to attach to the same scope).
- When the copy is for a new scope, the "update_symbols_from" (or "replace_symbols_with") can be used to replace the symbols using their name to the symbols in the new scope.

arporter commented on src/psyclone/psyir/nodes/literal.py:
```diff
@@ -171,3 +169,19 @@ def node_str(self, colour=True):
+        :type table: :py:class:`psyclone.psyir.symbols.SymbolTable`
+        '''
+        self.datatype.update_symbols_from(table)
+        # Literal is a leaf so no need to call super().
```
Yes, good point. It's true at the moment but may not be if something changes in future.

arporter commented on src/psyclone/psyir/symbols/datasymbol.py:
```diff
@@ -298,15 +298,30 @@ def copy(self):
         original will not be affected so the copy will not be referred
         to by any other object.
 
+        N.B. any Symbols referenced in either the datatype or the initial_value
+        properties will remain unchanged in the copied object. See the
```
Yes, done.

arporter commented on src/psyclone/psyir/symbols/datatypes.py:
```diff
@@ -158,7 +177,8 @@ def partial_datatype(self):
         :returns: partial datatype information if it can be determined,
             else None.
-        :rtype: Optional[:py:class:`psyclone.symbols.DataType`]
+        :rtype: NoneType | :py:class:`psyclone.psyir.symbols.DataType` |
+            :py:class:`psyclone.symbols.symbols.DataTypeSymbol`
```
Done. I can't remember who it was but someone didn't like `Optional` for a return value. (Might have been Rupert?)

arporter commented on src/psyclone/psyir/symbols/datatypes.py:
```diff
@@ -67,6 +68,25 @@ def __eq__(self, other):
 
+    def copy(self):
+        '''
+        :returns: a shallow copy of this datatype.
```
Done all.

arporter commented on src/psyclone/psyir/symbols/typed_symbol.py:
```diff
@@ -124,14 +124,15 @@ def copy(self):
         # first positional argument.
-        return type(self)(self.name, self.datatype, visibility=self.visibility,
+        return type(self)(self.name, self.datatype.copy(),
+                          visibility=self.visibility,
                           interface=self.interface)
```
Good spot, done. I had to implement a new `__eq__` for `ImportInterface` too.

arporter commented on src/psyclone/psyir/symbols/datatypes.py:
```diff
@@ -335,8 +380,33 @@ def __eq__(self, other):
 
-        return (self.precision == other.precision and
-                self.intrinsic == other.intrinsic)
+        if isinstance(other.precision, Symbol) and isinstance(self.precision,
+                                                              Symbol):
+            # If the precision in both types is given by a Symbol then we just
+            # compare the names of those symbols.
+            precision_match = other.precision.name == self.precision.name
+        else:
+            precision_match = self.precision == other.precision
+        return precision_match and self.intrinsic == other.intrinsic
```
One instance of the problem is in the kernel extraction tests:
```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> captured stdout >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
ARPDBG: Added i_def 140716787661664 to table of InvokeSchedule:
End InvokeSchedule

ARPDBG: Added i_def 140716783407840 to table of Routine[name:'testkern_coord_w0_2_code']:
End Routine
ARPDBG: Added i_def 140716783415616 to table of None
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
(Pdb) id(datatype.precision)
140716787661664
(Pdb) id(sym.datatype.precision)
140716783415616
```

arporter replied:
So we have two different `LFRicIntegerScalarDataType` instances, each containing a different 'i_def' symbol as its precision. Those two should be the *same* symbol.

arporter replied:
```
PSyclone/src/psyclone/domain/lfric/lfric_extract_driver_creator.py:932: in create
    child.lower_to_language_level()
PSyclone/src/psyclone/domain/lfric/lfric_loop.py:140: in lower_to_language_level
    start = self.start_expr.copy()
PSyclone/src/psyclone/domain/lfric/lfric_loop.py:780: in start_expr
    lbound = sym_table.find_or_create_integer_symbol(root_name,
```
It's the "new" `datatype` created in the last routine that has the reference to `i_def` in the original InvokeSchedule. However, at this point, we're dealing with a copy of that schedule (as we're in the process of making a new driver).

arporter replied:
I've create #2659 and added a TODO.

arporter replied:
All in all, that was another good spot so thanks. I've tightened up what I have so that it also checks that the interface of the two precision symbols matches.

arporter commented on src/psyclone/psyir/symbols/datasymbol.py:
```diff
@@ -327,3 +342,20 @@ def copy_properties(self, symbol_in):
+
+    def update_symbols_from(self, table):
+        '''
+        Replace any Symbols referred to by this object with those of the
+        same name in the supplied SymbolTable. If, for a given Symbol, there
+        is no corresponding entry in the supplied table, then that
+        Symbol is left unchanged.
```
Done.

arporter commented:
I need to update the tests for Loop to explicitly cover the new `replace_symbols_using` method.

arporter commented:
Failure is in LFRic adjoint example, `poly1d_w3_reconstruction_kernel_mod`. This has a loop with variable `stencil` and when we attempt to update this we are accidentally trying to use the symbol of the same name that is used in the metadata (i.e. imported from `argument_mod`). I need to see why the new symbol table doesn't have the local `stencil` definition in it.

sergisiso commented:
> Failure is in LFRic adjoint example, `poly1d_w3_reconstruction_kernel_mod`. This has a loop with variable `stencil` and when we attempt to update this we are accidentally trying to use the symbol of the same name that is used in the metadata (i.e. imported from `argument_mod`). I need to see why the new symbol table doesn't have the local `stencil` definition in it.
> 
> It's because it is the wrong symbol table - `replace_symbols_from` itself recurses and so is being called on Loop but with the symbol table of the parent scoping node.

I was expecting this to happen and was surprised that we didn't hit this sooner, I guess loop variables is the only use case where we use nested scope variables in practice (and it needs a clash to fail because not founds are not updated). But I think this is good because fixing it will  (I think) make trivial to solve #2424 

I agree the solution is to change the symbol table to the inner scope when reversing symbols. But currently you are doing it at _refine_copy which means that other calls to replace_symbols_using will still have this problem. I am wondering if this behaviour is expected always for `replace_symbols_using`? So, if we have:

```
subrotine test
   integer :: a
   integer :: b = 1

  if condition
      ! PSyIR declares a shadowed locally-scoped a'
      a' = 1
      if condition2
          ! PSyIR declares a shadowed locally-scoped b'
          b' = 2
          a = a' + b'
     end if
  end if
end subroutine test
```
and we do, `routine.replace_symbols_using(SymbolTable<contains a and b>)`
I would not expect this to change a' and b' to use the routine-scoped versions, as this will change the meaning of the code.

This can be done overriding the ScopingNode::replece_symbols_using, instead of modifying its _refine_copy 
```
def replace_symbols_using(self, table):
    for child in self.children:
         child.replace_symbols_using(self.symbol_table)
```
What do you think?

Note that in my example I cheekily added an statement containing both a and a'. Neither what I am proposing nor the current solution solves this. And a copy will change the semantics of this PSyIR tree! This case is rare, and the PR big enough, so it does not need to be solved here, but if it is not at least it should be documented.

arporter commented:
Finally, this is ready for another look. It's certainly nicer than it was before after the latest update so thanks for the suggestions.

sergisiso commented on src/psyclone/psyir/nodes/reference.py:
```diff
@@ -250,6 +250,41 @@ def next_access(self):
+        supplied table with matching names. If there is no match for a given
+        Symbol then it is left unchanged.
+
+        .. note:
+            Since we only use symbol names here, this won't get the correct
+            symbol if we have symbols shadowed in nested scopes, e.g.:
+
+            subroutine test
+              integer :: a
+              integer :: b = 1
+              if condition
+                ! PSyIR declares a shadowed locally-scoped a'
+                a' = 1
+                if condition2
+                  ! PSyIR declares a shadowed locally-scoped b'
+                  b' = 2
+                  a = a' + b'
```
Thanks for adding this, but it is not specific to reference, it will happen to any node with a symbol  inside a nested scope.  Could this note be in the ScopingNode implementation or  in the developer guide new section?

sergisiso replied:
And if fact it should probably be a note regarding the copy/_refine_copy instead, since we actually conform exactly to the replace_symbols_using doctring here.

The flaw is in the copy context as this is not exactly what we need there since it changes the semantics of the resulting node tree. I would also add a issue/TODO for this copy behaviour.

arporter replied:
Yes, agreed. I've moved it to the appropriate place, created #2666 and updated the dev. guide too.

sergisiso commented on src/psyclone/psyir/symbols/generic_interface_symbol.py:
```diff
@@ -147,9 +147,30 @@ def copy(self):
         # The constructors for all Symbol-based classes have 'name' as the
         # first positional argument.
-        return type(self)(self.name, self.routines, datatype=self.datatype,
+        return type(self)(self.name, self.routines[:],
```
Is the [:] needed?

arporter replied:
I did it to ensure we get a copy of the `routines` list otherwise the new class might have the same list as the old one.

sergisiso commented on src/psyclone/psyir/nodes/scoping_node.py:
```diff
@@ -96,29 +95,39 @@ def _refine_copy(self, other):
-                        node.variable.name)
+        # Now we've updated the symbol table, walk back down the tree and
+        # update any Symbols.
+        self.replace_symbols_using(self._symbol_table)
```
Maybe not a TODO, but I would add a comment saying that this replicates work since its called multiple times when going up the _refine_copy tree, so we may want to improve it in the future.

arporter replied:
Yes, note added.

sergisiso commented on src/psyclone/tests/psyir/frontend/fparser2_subroutine_handler_test.py:
```diff
@@ -207,11 +207,11 @@ def test_function_type_prefix(fortran_reader, fortran_writer,
         f"    {basic_type} :: my_fUnc\n"
         f"\n"
-        f"    my_fUnc = {rhs_val}\n"
+        f"    my_func = {rhs_val}\n"
```
Just to double check, are the changes in this test still necessary?

arporter replied:
No, I don't think they are.

sergisiso commented on doc/developer_guide/psyir_symbols.rst:
```diff
@@ -220,6 +216,41 @@ the `__contains__` method has no mechanism to pass a `scope_limit`
+
+.. automethod:: psyclone.psyir.symbols.Symbol.copy
+
+This ensures that the precision and initial-value properties are copied
+appropriately (e.g. new PSyIR nodes constructed) but any Symbols
+referred to by those properties remain unchanged.
+
+When performing a deep copy of a PSyIR tree, obviously all Symbols will
```
Maybe: "
This ensures that the precision and initial-value PSyIR subtrees are also copied appropriately, but any Symbols inside those nodes remain referring to the same Symbol objects (and therefore can still be used in the same scope).

However, when performing a deep copy of a PSyIR tree, all Symbols ..."

arporter replied:
I've re-worded. See what you think.

sergisiso commented on doc/developer_guide/psyir_symbols.rst:
```diff
@@ -220,6 +216,41 @@ the `__contains__` method has no mechanism to pass a `scope_limit`
+
+.. automethod:: psyclone.psyir.symbols.Symbol.replace_symbols_using
+
+All PSyIR `Node` classes also implement this method. In this case, it
```
Remove "also"?

arporter replied:
I'd like to keep it to differentiate from the previous text which is about `Symbol` - it's slightly unusual for two, distinct classes to have the same method name.

sergisiso reviewed (changes requested):
Fantastic work @arporter. There is final documentation suggestions but then this can go.

sergisiso commented:
Also check if the sphinx-link-check issue is relevant and launch an integration test since I see non of the last one where complete.

arporter commented:
Link check failure seems to have sorted itself out. I've triggered the integration tests again. Ready for another look, CI-permitting.

arporter commented:
LFRic 'everything-everywhere-all-at-once' integration test failed due to hitting max. recursion depth.
Problem occurs in backend when processing (inlined?) kernel `mm_diagonal_kernel_code_r_double` for `reconstruct_w3_field_alg_mod_psy`.

arporter commented:
The problem was that in `Reference.replace_symbols_using()` I had added a call to `replace_symbols_using` on the Symbol itself. This then hit an infinite recursion if the Symbol had an initial value defined by a PSyIR expression. Since it is the responsibility of the SymbolTable to update its symbols, I've simply removed the offending call and updated a failing test.

arporter commented:
Integration tests were green and coverage is all OK. Really ready for review...

sergisiso reviewed (approved):
@arporter All comments have been addressed, the documentation has been updated explaining the new internal details and it builds correctly. The integration passes (and the system was noisy but looking at the best bars on multiple runs it seems the performance is unaffected). This is ready to merge.
